### ISSUE TITLE
cant send any message

### ISSUE THREAD

**Issue by MrMontesa** [zikula-ev/Formicula#72]
Zikula: 1.4.4
Formicula: 4.0.0
I added permisson rule: unregistered ZikulaFormiculsModule:: :: access_comment

Unregistered users can now see form0 and select a contact. But when clicking "send" they are redirected to the login page.
For testing purpose I logged in with my admin account. Result was "Access denied" page. No form has been submitted.
While doing this I "tailed" prod.log & mailer.log and found there lines:

when submitting the form befor I got redicted:

```
[2017-01-16 13:29:59] app.ERROR: Event handler was not attached for event 'module.content.gettypes' because method 'AddressBook_EventHandler_Listeners::getContentTypes' is not callable [] []
[2017-01-16 13:29:59] request.INFO: Matched route "zikulaformiculamodule_user_index". {"route_parameters":{"_controller":"Zikula\\FormiculaModule\\Controller\\UserController::indexAction","_zkBundle":"ZikulaFormiculaModule","_zkModule":"ZikulaFormiculaModule","_zkType":"user","_zkFunc":"index","_locale":"en","_route":"zikulaformiculamodule_user_index"},"request_uri":"http://www.amc-idstein.de/formicula/"} []
[2017-01-16 13:29:59] app.DEBUG: The method signature for filter() has changed. See \Symfony\Component\HttpFoundation\ParameterBag::filter(). [] []
[2017-01-16 13:29:59] app.DEBUG: The method signature for filter() has changed. See \Symfony\Component\HttpFoundation\ParameterBag::filter(). [] []
[2017-01-16 13:29:59] app.ERROR: Event handler was not attached for event 'module.content.gettypes' because method 'AddressBook_EventHandler_Listeners::getContentTypes' is not callable [] []
[2017-01-16 13:30:00] request.INFO: Matched route "zikulausersmodule_access_login". {"route_parameters":{"returnUrl":"%2Fformicula%2F","_controller":"Zikula\\UsersModule\\Controller\\AccessController::loginAction","_zkBundle":"ZikulaUsersModule","_zkModule":"ZikulaUsersModule","_zkType":"access","_zkFunc":"login","_locale":"en","_route":"zikulausersmodule_access_login"},"request_uri":"http://www.amc-idstein.de/login/%252Fformicula%252F"} []
[2017-01-16 13:30:00] app.DEBUG: Warning! Class Zikula_Hook_AbstractHandler is deprecated. [] []
[2017-01-16 13:30:00] app.ERROR: Event handler was not attached for event 'module.content.gettypes' because method 'AddressBook_EventHandler_Listeners::getContentTypes' is not callable [] []
[2017-01-16 13:30:01] request.INFO: Matched route "legacy_short_url". {"route_parameters":{"_controller":"zikula_core.controller.legacy_controller:shortUrlAction","path":"favicon.ico","_locale":"en","_route":"legacy_short_url"},"request_uri":"http://www.amc-idstein.de/favicon.ico"} []
[2017-01-16 13:30:01] request.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "Page not found." at /data/web/1/000/027/915/75642/htdocs/amc-idstein/lib/Zikula/Bundle/CoreBundle/Controller/LegacyController.php line 128 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): Page not found. at /data/web/1/000/027/915/75642/htdocs/amc-idstein/lib/Zikula/Bundle/CoreBundle/Controller/LegacyController.php:128)"} []
[2017-01-16 13:30:10] app.ERROR: Event handler was not attached for event 'module.content.gettypes' because method 'AddressBook_EventHandler_Listeners::getContentTypes' is not callable [] []
```

After loggin in receiving the access denied error message:

```
[2017-01-16 13:30:15] app.ERROR: Event handler was not attached for event 'module.content.gettypes' because method 'AddressBook_EventHandler_Listeners::getContentTypes' is not callable [] []
[2017-01-16 13:30:15] request.INFO: Matched route "zikulausersmodule_access_login". {"route_parameters":{"returnUrl":"%2Fformicula%2F","_controller":"Zikula\\UsersModule\\Controller\\AccessController::loginAction","_zkBundle":"ZikulaUsersModule","_zkModule":"ZikulaUsersModule","_zkType":"access","_zkFunc":"login","_locale":"en","_route":"zikulausersmodule_access_login"},"request_uri":"http://www.amc-idstein.de/login/%252Fformicula%252F"} []
[2017-01-16 13:30:15] app.DEBUG: Warning! Class Zikula_Hook_AbstractHandler is deprecated. [] []
[2017-01-16 13:30:15] app.ERROR: Event handler was not attached for event 'module.content.gettypes' because method 'AddressBook_EventHandler_Listeners::getContentTypes' is not callable [] []
[2017-01-16 13:30:16] request.INFO: Matched route "zikulaformiculamodule_user_index". {"route_parameters":{"_controller":"Zikula\\FormiculaModule\\Controller\\UserController::indexAction","_zkBundle":"ZikulaFormiculaModule","_zkModule":"ZikulaFormiculaModule","_zkType":"user","_zkFunc":"index","_locale":"en","_route":"zikulaformiculamodule_user_index"},"request_uri":"http://www.amc-idstein.de/formicula/"} []
[2017-01-16 13:30:16] app.DEBUG: The method signature for filter() has changed. See \Symfony\Component\HttpFoundation\ParameterBag::filter(). [] []
[2017-01-16 13:30:16] app.DEBUG: The method signature for filter() has changed. See \Symfony\Component\HttpFoundation\ParameterBag::filter(). [] []
[2017-01-16 13:30:16] app.DEBUG: Warning! Class Zikula_Hook_AbstractHandler is deprecated. [] []
[2017-01-16 13:30:17] app.ERROR: Event handler was not attached for event 'module.content.gettypes' because method 'AddressBook_EventHandler_Listeners::getContentTypes' is not callable [] []
[2017-01-16 13:30:17] request.INFO: Matched route "legacy_short_url". {"route_parameters":{"_controller":"zikula_core.controller.legacy_controller:shortUrlAction","path":"favicon.ico","_locale":"en","_route":"legacy_short_url"},"request_uri":"http://www.amc-idstein.de/favicon.ico"} []
[2017-01-16 13:30:17] request.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "Page not found." at /data/web/1/000/027/915/75642/htdocs/amc-idstein/lib/Zikula/Bundle/CoreBundle/Controller/LegacyController.php line 128 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): Page not found. at /data/web/1/000/027/915/75642/htdocs/amc-idstein/lib/Zikula/Bundle/CoreBundle/Controller/LegacyController.php:128)"} []
```

I do have Addressbook and Content installed along with Formicula but cant imagine how this interfears with each other,

Edit: Mailing in general works fine as I do receive "new user registered" emails, thus I suspect mailer module is ok.

**Comment by Guite** [comment]
The error message refers to the AddressBook module. Thus, you should try to disable it and see if it works then.

**Comment by MrMontesa** [comment]
Still Access denies error message, but log output is shorter:

> [2017-01-16 13:44:55] request.INFO: Matched route "legacy_short_url". {"route_parameters":{"_controller":"zikula_core.controller.legacy_controller:shortUrlAction","path":"favicon.ico","_locale":"en","_route":"legacy_short_url"},"request_uri":"http://www.amc-idstein.de/favicon.ico"} []
> [2017-01-16 13:44:55] request.ERROR: Uncaught PHP Exception Symfony\Component\HttpKernel\Exception\NotFoundHttpException: "Page not found." at /data/web/1/000/027/915/75642/htdocs/amc-idstein/lib/Zikula/Bundle/CoreBundle/Controller/LegacyController.php line 128 {"exception":"[object] (Symfony\\Component\\HttpKernel\\Exception\\NotFoundHttpException(code: 0): Page not found. at /data/web/1/000/027/915/75642/htdocs/amc-idstein/lib/Zikula/Bundle/CoreBundle/Controller/LegacyController.php:128)"} []
> 
>

**Comment by Guite** [comment]
> I added permisson rule: unregistered ZikulaFormiculsModule:: :: access_comment

did you copy that from your permission table? Because there is a typo in the module name.

**Comment by MrMontesa** [comment]
No I dont. Before adding the rule, access was forbidden immedialty on accessing the page. After adding the rule, I was able to see the form. So I think the rule is working ok. Also tried higher permissions (access_edit) without luck. Even my admin user receives access denied, which is a bit wired.
Heres a c&p from the permission rule:

```
Unregistered
ZikulaFormiculaModule::
::
access_comment
```

**Comment by paustian** [comment]
Guite, I think I tracked this error down. If you look at lines 139 to 144 of the code in Controller/UserController.php. 
`$formid = $formData['form'];
            $contactId = $formData['cid'];
            if (empty($contactId) || empty($formId) || !$this->hasPermission('ZikulaFormiculaModule::', "$formId:$contactId:", ACCESS_COMMENT)) {
                throw new AccessDeniedException();
            }`
$formid should be $formId. Now this still fails because with the 0 form, $formID set to 0 is interpreted by php as being false. If you look at the docs for empty in the php manual

> Determine whether a variable is considered to be empty. A variable is considered empty if it does not exist or if its value equals FALSE. empty() does not generate a warning if the variable does not exist.

So this still fails. A fix for this is to use formId<0? That will work.

**Comment by MrMontesa** [comment]
Just patched my system with the 2 lines from the patch. now my page bails out with error 500. Full stack trace:

```
Attempted to call an undefined method named "dispatchHooks" of class "Zikula\FormiculaModule\Controller\UserController".
in modules/Formicula/Controller/UserController.php at line 207   -
            // Check hooked modules for validation
            $validationHook = new ValidationHook(new ValidationProviders());
            $validators = $this->dispatchHooks('formicula.ui_hooks.forms.validate_edit', $validationHook)->getValidators();
            if ($validators->hasErrors()) {
                $this->addFlash('error', $this->__('The validation of the hooked security module was incorrect. Please try again.'));
                $hasError = true;
```

**Comment by Guite** [comment]
Fixed in bfd8ea93bc40a29857d366759de793ad6d6cba83
In future please create new issues for new problems. Thanks.
